lib/SILOptimizer/Transforms/VTableSpecializer.cpp

```diff
@@ -0,0 +1,222 @@
+
+  if (!ReInfo.canBeSpecialized()) {
+    llvm::errs() << "Cannot specialize vtable method " << origMethod->getName()
+                 << '\n';
+    llvm::report_fatal_error("cannot specialize vtable method");
+  }
+
+  SILOptFunctionBuilder FunctionBuilder(*this);
+
+  GenericFuncSpecializer FuncSpecializer(FunctionBuilder, origMethod, subs,
+                                         ReInfo, /*isMandatory=*/true);
+  SILFunction *SpecializedF = FuncSpecializer.lookupSpecialization();
+  if (!SpecializedF) SpecializedF = FuncSpecializer.tryCreateSpecialization();
+  if (!SpecializedF || SpecializedF->getLoweredFunctionType()->hasError()) {
+    llvm::errs()
+        << "Cannot specialize vtable method " << origMethod->getName() << '\n'
+        << "Generic class methods are not supported in embedded mode\n";
+    llvm::report_fatal_error("cannot specialize vtable method");
+  }
+
+  // Link after specializing to pull in everything referenced from another
+  // module in case some referenced functions have non-public linkage.
+  module.linkFunction(SpecializedF, SILModule::LinkingMode::LinkAll);
+
+  SpecializedF->setLinkage(SILLinkage::Public);
+  SpecializedF->setSerialized(IsNotSerialized);
+
+  return SpecializedF;
+}
+
+bool VTableSpecializer::specializeClassMethodInst(ClassMethodInst *cm) {
+  SILValue instance = cm->getOperand();
+  SILType classTy = instance->getType();
+  CanType astType = classTy.getASTType();
+  BoundGenericClassType *genClassTy = dyn_cast<BoundGenericClassType>(astType);
+  if (!genClassTy) return false;
+
+  ClassDecl *classDecl = genClassTy->getDecl();
+  SubstitutionMap subs = astType->getContextSubstitutionMap(
+      classDecl->getParentModule(), classDecl);
+
+  SILType funcTy = cm->getType();
+
+  SILFunction *f = cm->getFunction();
+  SILModule &m = f->getModule();
+  SILType substitutedType =
+      funcTy.substGenericArgs(m, subs, TypeExpansionContext::minimal());
+
+  ReabstractionInfo reInfo(substitutedType.getAs<SILFunctionType>(), m);
+  reInfo.createSubstitutedAndSpecializedTypes();
+  CanSILFunctionType finalFuncTy = reInfo.getSpecializedType();
+  SILType finalSILTy = SILType::getPrimitiveObjectType(finalFuncTy);
+
+  SILBuilder builder(cm);
+  auto *newCM = builder.createClassMethod(cm->getLoc(), cm->getOperand(),
+                                          cm->getMember(), finalSILTy);
+
+  while (!cm->use_empty()) {
+    Operand *use = *cm->use_begin();
+    SILInstruction *user = use->getUser();
+    ApplySite AI = ApplySite::isa(user);
+    if (AI && AI.getCalleeOperand() == use) {
+      replaceWithSpecializedCallee(AI, newCM, reInfo);
+      AI.getInstruction()->eraseFromParent();
+      continue;
+    }
+    llvm::errs() << "unsupported use of class method "
+                 << newCM->getMember().getDecl()->getName() << " in function "
+                 << newCM->getFunction()->getName() << '\n';
+    llvm::report_fatal_error("unsupported class method");
```

Comment:
What's the case where this could happen? If there are uses of a class method that aren't an apply? For example escaping the function pointer maybe? Seems like some of these fatal errors should be turned into diagnostics (that can be a follow up though).